powdr/src/lib.rs

```diff
@@ -64,25 +65,53 @@ impl Session {
     }
 
     pub fn prove(&mut self) {
-        match self.pipeline.read_constants_mut(Path::new(&self.out_path)) {
-            Ok(_) => (),
-            Err(_) => {
-                self.pipeline.compute_fixed_cols().unwrap();
+        let asm_name = self.pipeline.asm_string().unwrap().0.clone().unwrap();
+        let pil_file = create_pil_file_path(&asm_name);
+
+        let generate_artifacts = if let Some(existing_pil) = read_file_if_exists(&pil_file) {
+            let computed_pil = self.pipeline.compute_optimized_pil().unwrap().to_string();
+            if existing_pil != computed_pil {
+                log::info!("Compiled PIL changed, invalidating artifacts...");
+                true
+            } else {
+                log::info!("Compiled PIL did not change, will try to reuse artifacts...");
+                false
             }
-        }
+        } else {
+            log::info!("PIL file not found, will generate artifacts...");
+            true
+        };
 
         let pkey = Path::new(&self.out_path).join(DEFAULT_PKEY);
         let vkey = Path::new(&self.out_path).join(DEFAULT_VKEY);
 
-        if pkey.exists() && vkey.exists() {
-            self.pipeline.set_pkey_file(pkey.clone());
-            self.pipeline.set_vkey_file(vkey.clone());
-            self.pipeline.setup_backend().unwrap();
-        } else {
+        if generate_artifacts {
+            self.pipeline.compute_fixed_cols().unwrap();
             self.pipeline.setup_backend().unwrap();
             self.export_setup();
             self.pipeline.set_pkey_file(pkey.clone());
```

Comment:
this `set_pkey_file` and `set_vkey_file` are done in all if/else paths, can they be outside?

Reply:
The annoying part here is whether they're before or after `self.pipeline.setup_backend()`. If they exist, we want them to be before setup_backend so they're loaded during setup. If they don't, we want to do the setup before (which writes them to disk) and then set them to pipeline in case it gets cloned etc. I'll check if the latter can be done inside pipeline.

Reply:
Yea I think we have to set it after for now, otherwise it's gonna try to get the keys from non-existing or empty files